File: ts-packages/web/src/features/spaces/rewards/types/reward-condition.ts
export type RewardCondition =
  | 'None'
  | { MaxClaims: number }
  | { MaxPoints: number }
  | { MaxUserClaims: number }
  | { MaxUserPoints: number };

export enum ConditionType {
  None = 'None',
  MaxClaims = 'MaxClaims',
  MaxPoints = 'MaxPoints',
  MaxUserClaims = 'MaxUserClaims',
  MaxUserPoints = 'MaxUserPoints',
}
export function isConditionNone(c: RewardCondition): c is 'None' {
  return c === 'None';
}

export function getConditionValue(c: RewardCondition): number | null {
  if (c === 'None') return null;
  if ('MaxClaims' in c) return c.MaxClaims;
  if ('MaxPoints' in c) return c.MaxPoints;
  if ('MaxUserClaims' in c) return c.MaxUserClaims;
  if ('MaxUserPoints' in c) return c.MaxUserPoints;
  return null;
}

export function getConditionType(c: RewardCondition): ConditionType {
  if (c === 'None') return ConditionType.None;
  return Object.keys(c)[0] as ConditionType;
}

File: packages/main-api/src/features/spaces/rewards/types/reward_condition.rs
use crate::*;
use chrono::Datelike;

#[derive(
    Debug,
    Default,
    Clone,
    Serialize,
    Deserialize,
    JsonSchema,
    OperationIo,
    DynamoEnum,
    Eq,
    PartialEq,
)]
pub enum RewardCondition {
    #[default]
    None,
    // 전체 Reward 최대 횟 수 제한
    MaxClaims(i64),
    // 전체 Reward 최대 포인트 제한
    MaxPoints(i64),
    // 유저 당 Reward 수 제한
    MaxUserClaims(i64),
    // 유저 당 포인트 제한
    MaxUserPoints(i64),
}
